numalogic/registry/redis_registry.py

```diff
@@ -81,24 +84,71 @@ def get_version(key: str) -> str:
         """
         return key.split("::")[-1]
 
-    def __get_model_key(self, latest: bool, version: str, key: str) -> str:
-        if latest:
-            production_key = self.__construct_production_key(key)
-            if not self.client.exists(production_key):
-                raise ModelKeyNotFound(
-                    f"Production key: {production_key}, Not Found !!!\n Exiting....."
-                )
-            model_key = self.client.get(production_key)
-            if not self.client.exists(model_key):
-                raise ModelKeyNotFound(
-                    "Production key = {} is pointing to the key: {} that "
-                    "is missing the redis registry".format(production_key, model_key)
-                )
-        else:
-            model_key = self.__construct_version_key(key, version)
-            if not self.client.exists(model_key):
-                raise ModelKeyNotFound("Could not find model key with key: %s" % model_key)
-        return model_key
+    def _load_from_cache(self, key: str) -> Optional[ArtifactData]:
+        if not self.cache_registry:
+            return None
+        return self.cache_registry.load(key)
+
+    def _save_in_cache(self, key: str, artifact_data: ArtifactData) -> None:
+        if self.cache_registry:
+            self.cache_registry.save(key, artifact_data)
+
+    def _clear_cache(self, key: Optional[str] = None) -> Optional[ArtifactData]:
+        if self.cache_registry:
+            if key:
+                return self.cache_registry.delete(key)
+            return self.cache_registry.clear()
+        return None
+
+    def __get_artifact_data(
+        self,
+        model_key: str,
+    ) -> ArtifactData:
+        (
+            serialized_artifact,
+            artifact_version,
+            artifact_timestamp,
+            serialized_metadata,
+        ) = self.client.hmget(name=model_key, keys=["artifact", "version", "timestamp", "metadata"])
+        deserialized_artifact = loads(serialized_artifact)
+        deserialized_metadata = None
+        if serialized_metadata:
+            deserialized_metadata = loads(serialized_metadata)
+        return ArtifactData(
+            artifact=deserialized_artifact,
+            metadata=deserialized_metadata,
+            extras={
+                "timestamp": float(artifact_timestamp.decode()),
+                "version": artifact_version.decode(),
+            },
+        )
+
+    def __load_latest_artifact(self, key: str) -> ArtifactData:
+        cached_artifact = self._load_from_cache(key)
+        if cached_artifact:
+            return cached_artifact
+        production_key = self.__construct_production_key(key)
+        if not self.client.exists(production_key):
+            raise ModelKeyNotFound(
+                f"Production key: {production_key}, Not Found !!!\n Exiting....."
+            )
+        model_key = self.client.get(production_key)
```

Comment:
why not call `__load_version_artifact` here?

Reply:
Actually a good idea!! Let me do that and another logger statement for it.